mathics/builtin/assignment.py

```diff
@@ -65,376 +74,577 @@ def get_symbol_list(list, error_callback):
+            raise AssignmentException(lhs, None)
+        for leaf in focus.leaves:
+            name = leaf.get_lookup_name()
+            tags.append(name)
+    else:
+        allowed_names = [focus.get_lookup_name()]
+        for leaf in focus.get_leaves():
+            if not leaf.is_symbol() and leaf.get_head_name() in ("System`HoldPattern",):
+                leaf = leaf.leaves[0]
+            if not leaf.is_symbol() and leaf.get_head_name() in ("System`Pattern",):
+                leaf = leaf.leaves[1]
+            if not leaf.is_symbol() and leaf.get_head_name() in (
+                "System`Blank",
+                "System`BlankSequence",
+                "System`BlankNullSequence",
             ):
-                evaluation.message(lhs_name, "cxset", rhs)
-                return False
+                if len(leaf.leaves) == 1:
+                    leaf = leaf.leaves[0]
+
+            allowed_names.append(leaf.get_lookup_name())
+        for name in tags:
+            if name not in allowed_names:
+                evaluation.message(self.get_name(), "tagnfd", Symbol(name))
+                raise AssignmentException(lhs, None)
+
+    return tags, focus
+
+
+# Here starts the functions that implement `assign_elementary` for different
+# kind of expressions. Maybe they should be put in a separated module or
+# maybe they should be member functions of _SetOperator.
+
+
+def process_assign_messagename(self, lhs, rhs, evaluation, tags, upset):
+    lhs, condition = unroll_conditions(lhs)
+    lhs, rhs = unroll_patterns(lhs, rhs, evaluation)
+    count = 0
+    defs = evaluation.definitions
+    if len(lhs.leaves) != 2:
+        evaluation.message_args("MessageName", len(lhs.leaves), 2)
+        raise AssignmentException(lhs, None)
+    focus = lhs.leaves[0]
+    tags = process_tags_and_upset_dont_allow_custom(
+        tags, upset, self, lhs, focus, evaluation
+    )
+    lhs, rhs = process_rhs_conditions(lhs, rhs, condition, evaluation)
+    rule = Rule(lhs, rhs)
+    for tag in tags:
+        if rejected_because_protected(self, lhs, tag, evaluation):
+            continue
+        count += 1
+        defs.add_message(tag, rule)
+    return count > 0
+
+
+def process_assign_default(self, lhs, rhs, evaluation, tags, upset):
+    lhs, condition = unroll_conditions(lhs)
+    lhs, rhs = unroll_patterns(lhs, rhs, evaluation)
+    count = 0
+    defs = evaluation.definitions
+
+    if len(lhs.leaves) not in (1, 2, 3):
+        evaluation.message_args("Default", len(lhs.leaves), 1, 2, 3)
+        raise AssignmentException(lhs, None)
+    focus = lhs.leaves[0]
+    tags = process_tags_and_upset_dont_allow_custom(
+        tags, upset, self, lhs, focus, evaluation
+    )
+    lhs, rhs = process_rhs_conditions(lhs, rhs, condition, evaluation)
+    rule = Rule(lhs, rhs)
+    for tag in tags:
+        if rejected_because_protected(self, lhs, tag, evaluation):
+            continue
+        count += 1
```

Comment:
I believe this isn't necessary. We could instead return `len(tags) > 0`.

Reply:
@TiagoCavalcante, that does not work: it is possible that all the tags are rejected because of symbol protection...

Reply:
instead of count, we can just set a flag to True, I do not know what is the best.

Reply:
When we are at the level of haggling over to use a boolean or an integer count, I suspect we are focusing on stuff that isn't going to matter that much. 

Yes, one is faster than the other, but not by that much to make a difference. Of course if you want to benchmark this and you come up with more than a one or two percent difference - great! Prove me wrong!

Reply:
@rocky, I do not say it for performance, but for clarity: I had to spend some time trying to figure out why we need a number, which in the end is not used.

Reply:
@mmatera @rocky I believe here a boolean is better.

Reply:
> @rocky, I do not say it for performance, but for clarity: I had to spend some time trying to figure out why we need the number of modified tags, which in the end is not used. And for the @TiagoCavalcante`s comment, it happened the same to him.

Good - thanks for the clarification. I greatly appreciate the effort spent to make things more understandable. Thanks.